### DISCUSSION TITLE
Advanced printing options for LBP2900B (toner density)

### DISCUSSION THREAD

**Discussion by apoorvpotnis** [mounaiban/captdriver#24] [Q&A] [answered]
Is there a way to access advanced printing options like toner density, collate, halftone patterns, image refinement, etc. with this driver? The only options which I see on `localhost:631/admin/#General` are media size, toner save, manual duplex, and media type. The `capt-src` driver has the other options but I am not able to install it (on Manjaro KDE) as python2 tests fail for some reason. `python2` is marked as a dependency for this driver (when installing from AUR). Thanks for this driver, btw.

**Comment by mounaiban** [comment, marked as answer]
From what I know about how CUPS and captdriver work, the only aspects of printing that we can control entirely from the printer is toner density and media type.

Manual duplex can be implemented by waiting on the user after every page from the driver. I think this was implemented since 104f414.

Media size is a shared responsibility between CUPS and captdriver; the size of the raster output by Ghostscript via CUPS must match (or be close enough) to what the printer is expecting.

Halftone patterns are handled by the raster image processor (RIP) on the host, I do not know any CAPT device with a built-in RIP like  high-end enterprise devices do... 

Which brings us to image refinement, which is still a mystery. Enabling it doesn't seem to have any meaningful effect on the output for now, but I believe this feature is enabled only with a custom RIP that can output specific halftone patterns that are recognised by the device, which are in turn output as finer details on the printouts.

I hope the driver solved your problem despite its incompleteness; sorry for the late reply...

**Reply by apoorvpotnis** [reply]
I see. Thank you so much for your answer and the driver. It was super helpful to me as I don't want to boot Windows just to print some documents. Printing docs is essential for me and this driver just saves the day. Thank you again.
